internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+// to advance the current reconcile, and that reconciliation should be retried.
+type reconcileError struct {
+	err error
+}
+
+func (r reconcileError) Result() (result reconcile.Result, err error) {
+	err = r.err
+	return
+}
+
+func (r reconcileError) Dirty() bool {
+	return false
+}
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;update;patch
+
+func (r *BMACReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		return reconcileComplete{}.Result()
+	}
+
+	// Let's reconcile the BMH
+	result := r.reconcileBMH(bmh)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.Errorf("Error updating hardwaredetails %v", err)
+			return reconcileError{err}.Result()
+		}
+
+		return result.Result()
+	}
+
+	// Should we check the status of the BMH resource?
+	// Provisioning / Deprovisioning ?
+	// What happens if we do the reconcile on a BMH that
+	// is in a Deprovisioning state?
+
+	agent := r.findAgent(ctx, bmh)
+
+	// handle multiple agents matching the
+	// same BMH's Mac Address
+
+	if agent == nil {
+		return result.Result()
+	}
+
+	// if we get to this point, it means we have found an Agent that is associated
+	// with the BMH being reconciled. We will call both, reconcileAgentSpec and
+	// reconcileAgentInventory, every time. The logic to decide whether there's
+	// any action to take is implemented in each functoin respectively.
+	result = r.reconcileAgentSpec(bmh, agent)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, agent)
+		if err != nil {
+			r.Log.Errorf("Error updating agent %v", err)
+			return reconcileError{err}.Result()
+		}
+
+		return result.Result()
+	}
+
+	result = r.reconcileAgentInventory(bmh, agent)
+	if result.Dirty() {
+		err := r.Client.Update(ctx, bmh)
+		if err != nil {
+			r.Log.Errorf("Error updating hardwaredetails %v", err)
+			return reconcileError{err}.Result()
+		}
+	}
+
+	return result.Result()
+}
+
+// This reconcile step takes care of copying data from the BMH into the Agent.
+//
+// The Agent's spec reconcile will only happen for non-Approved agents. These are
+// considered to be `new` and, at this point, the agent has not started the deployment
+// yet.
+//
+// We are only interested in the following data:
+//
+// - Agent role: bmac.adi.openshift.io/role
+// - Hostname: bmac.adi.openshift.io/hostname
+// - Machine Config Pool: bmac.adi.openshift.io/machine-config-pool
+//
+// Unless there are errors, the agent should be `Approved` at the end of this
+// reconcile and a label should be set on it referencing the BMH. No changes to
+// the BMH should happen in this reconcile step.
+func (r *BMACReconciler) reconcileAgentSpec(bmh *bmh_v1alpha1.BareMetalHost, agent *adiiov1alpha1.Agent) reconcileResult {
+
+	if agent.Spec.Approved {
+		return reconcileComplete{}
```

Comment:
It's possible that other values on the BMH could change post-approval, and I can't think of a reason not to propagate those changes. I think it would be better to go through all the values that get set rather than assume they all match based on this one value.